File: c-stdio/unisig.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "unisig.h"

char *read_unisig(FILE *input, int *out_is_uuid)
{
    const unsigned char magic[7]
        = { 0xdc, 0xdc, 0x0d, 0x0a, 0x1a, 0x0a, 0x00 };
    unsigned char head[sizeof(magic) + 1];
    char *sig;
    size_t len;
    int is_uuid;

    if (fread(head, 1, sizeof(head), input) != sizeof(head))
        return 0;
    if (memcmp(head, magic, sizeof(magic)))
        return 0;
    len = head[sizeof(magic)];
    *out_is_uuid = is_uuid = (len == 0);
    if (is_uuid)
        len = 16;
    sig = calloc(len + 1, 1);
    if (sig == NULL)
        goto fail; // Out of memory
    if (fread(sig, 1, len, input) != len)
        goto fail; // Read error or truncated file
    if (!is_uuid) {
        if (memchr(sig, '\0', len))
            goto fail; // Sig contains null byte, not usable as a C string.
    }
    return sig;
fail:
    free(sig);
    return NULL;
}

File: python-3/unisig.py
MAGIC = b"\xDC\xDC\x0D\x0A\x1A\x0A\x00"


def read(stream):
    head = stream.read(len(MAGIC) + 1)
    if len(head) != len(MAGIC) + 1 or head[: len(MAGIC)] != MAGIC:
        return False
    lenbyte = head[len(MAGIC)]
    is_uuid = lenbyte == 0
    length = 16 if is_uuid else lenbyte
    sig = stream.read(length)
    if len(sig) != length:
        return False
    return sig if is_uuid else sig.decode("utf-8", errors="strict")


def write(stream, name):
    if isinstance(name, bytes):
        assert len(name) == 16
        sig = name
        is_uuid = True
    else:
        assert isinstance(name, str)
        sig = name.encode("utf-8", errors="strict")
        is_uuid = False
    stream.write(MAGIC)
    stream.write(bytes([0 if is_uuid else len(sig)]))
    stream.write(sig)
